Initial Cinematic Dialog System

AndrewJakubowicz opened the pull request:
### tl;dr:
Major new file is DialogSystem.js.
Other changes are minor housekeeping changes and improvements.

## Simplified data transformation to get text on the screen.

![DialogSystem](https://user-images.githubusercontent.com/15080861/56757957-50a8ee80-674a-11e9-82ad-010f96e5e6d5.png)

## Features:

  - `/cinematic` route now redirects to `/` if not an admin.
  - Positioning text bubble
  - A simple template language leveraging a [super light library](https://blueimp.github.io/JavaScript-Templates/) that we've [forked and simplified](https://github.com/codecombat/JavaScript-Templates/tree/remove-load).
  - Localization added to schema and dialog bubbles correctly handle language.
  - Left and Right svg speaker bubbles. (svg programmer art)
  - Some light tests on the most important pieces. (text transpilation and transformation)
  - Clear all current text bubbles command.
  - All systems are now iterated for `parseDialogNode` per dialogNode.
  - All systems are now iterated for `parseSetupShot` per setupShot node.
  - i18n changes added to cinematic schema.
  - moved some logic into more rational places. Commands are now in a single file, and systems handle creating their own commands.

AndrewJakubowicz commented on app/views/play/cinematic/Command/AbstractCommand.js:
```diff
@@ -1,3 +1,4 @@
+// Throws and error if `import ... from ..` syntax.
```
```suggestion
// Throws an error if `import ... from ..` syntax.
```

AndrewJakubowicz commented on app/views/play/cinematic/CinematicLankBoss.js:
```diff
@@ -1,28 +1,63 @@
+   * @param {Shot} shot - the cinematic shot data.
+   */
+  parseSetupShot (shot) {
+    const leftCharSlug = getLeftCharacterThangTypeSlug(shot)
```
The code in this method was a move with no logic changes.

AndrewJakubowicz commented on app/views/play/cinematic/CinematicLankBoss.js:
```diff
@@ -128,40 +165,6 @@ export default class CinematicLankBoss {
-/**
- * AnimeCommand is used to turn animejs animation tweens into commands that the Command Runner can play and cancel.
- */
-class AnimeCommand extends AbstractCommand {
```
This code was also moved with no logic changes.

AndrewJakubowicz commented on app/views/play/cinematic/Command/AbstractCommand.js:
```diff
@@ -103,3 +104,54 @@ export class Noop extends AbstractCommand {
+/**
+ * AnimeCommand is used to turn animejs animation tweens into commands that the Command Runner can play and cancel.
+ */
+export class AnimeCommand extends AbstractCommand {
```
`AnimeCommand` was moved here with no logic changes.

AndrewJakubowicz commented on app/views/play/cinematic/Command/CinematicParser.js:
```diff
@@ -1,8 +1,40 @@
-import { getLeftCharacterThangTypeSlug, getRightCharacterThangTypeSlug } from '../../../../schemas/selectors/cinematic'
+import AbstractCommand from './AbstractCommand'
```
This file has logic changes.
We iterate through systems checking if they have particular setup methods.
If they do we give them cinematic data and concatenate a list of commands returned from all systems.

shubhi1092 commented on app/views/CinematicCanvas.vue:
```diff
@@ -28,6 +28,10 @@ export default {
     enterDisabled: true
   }),
   mounted: function() {
+    if (!me.isAdmin()) {
```
Alternatively, we can create a feature flag for this and do the admin check there like this: https://github.com/codecombat/codecombat/blob/master/app/models/User.coffee#L584

AndrewJakubowicz replied:
That's a brilliant idea.

shubhi1092 commented on app/views/play/cinematic/CinematicLankBoss.js:
```diff
@@ -1,28 +1,63 @@
 
 const Lank = require('lib/surface/Lank')
 
-Promise.config({
```
Why are we removing this

AndrewJakubowicz replied:
We don't need the cancellable promise, as it was being used in AnimeCommand.

shubhi1092 commented on app/views/play/cinematic/CinematicLankBoss.js:
```diff
@@ -68,8 +103,10 @@ export default class CinematicLankBoss {
       autoplay: false,
-      // Update required to ensure Lank sprite position is moved.
-      update: () => { this[side].thang.stateChanged = true }
+      easing: 'easeInOutQuart',
```
What is this for?

AndrewJakubowicz replied:
The default animation ease function has bounce which makes the characters bounce around a little bit too much. This is a much more pleasing easing function.

shubhi1092 commented on app/views/play/cinematic/Command/AbstractCommand.js:
```diff
@@ -103,3 +104,54 @@ export class Noop extends AbstractCommand {
+/**
+ * AnimeCommand is used to turn animejs animation tweens into commands that the Command Runner can play and cancel.
+ */
+export class AnimeCommand extends AbstractCommand {
```
Why did we move this from LankBoss to here?

AndrewJakubowicz replied:
Because we're able to reuse this Command for the dialogSystem.

shubhi1092 commented on app/views/play/cinematic/Command/CinematicParser.js:
```diff
@@ -13,33 +45,57 @@ export const parseShot = (shot, systems) => {
+  Object.values(systems)
+    .filter(sys => sys !== undefined && typeof sys.parseDialogNode === 'function')
+    .reduce((commands, sys) => (
+      [...commands, ...sys.parseDialogNode(dialogNode)]
```
Do we not need to check if `sys.parseDialogNode(dialogNode)` returns an array similar to `parseSetup` above?

AndrewJakubowicz replied:
Great comment. I had apparently not done this in this PR but include it in the next PR.
Thanks for this excellent pickup.

shubhi1092 commented on app/views/play/cinematic/cinematicController.js:
```diff
@@ -53,14 +54,25 @@ export class CinematicController {
     })
 
-    this.systems.loader = new Loader({ slug })
+    this.commands = []
```
Duplicate of line 77 below

AndrewJakubowicz replied:
Good catch

shubhi1092 commented on app/views/play/cinematic/Command/CinematicParser.js:
```diff
@@ -13,33 +45,57 @@ export const parseShot = (shot, systems) => {
+ * @returns {AbstractCommand[][]} A 2d array of commands.
+ */
+export const parseShot = (shot, systems) => {
+  const setupCommands = parseSetup(shot, systems) || []
```
Instead of sending all the systems to `parseSetup` and `parseDialogNode`, and then filtering there, I think it would be cleaner to filter here and then send only the specific system to `parseSetup` and `parseDialogNode`

shubhi1092 replied:
Can also probably rename `systems` to `cinematicSystems`?

AndrewJakubowicz replied:
For now I will not change  the filtering (as I think I might change this a little more in the next PR). But I did make  the name more clear. Thanks!

shubhi1092 commented on app/views/play/cinematic/DialogSystem.js:
```diff
@@ -0,0 +1,314 @@
+      y: y - height,
+      width,
+      height,
+      clazz: this.id,
```
Should it be `class` ?

AndrewJakubowicz replied:
Yeah this was some bad naming. `class` is a reserved keyword so I can't use that. `clazz` is not an easy name. It's flat out confusing. Great pickup.

shubhi1092 replied:
Oh i see, can we use `className` in that case.

AndrewJakubowicz replied:
That's such a better name.

shubhi1092 commented on app/views/play/cinematic/DialogSystem.js:
```diff
@@ -0,0 +1,314 @@
+ * @param {SvgBubbleOptions} svgOptions - Svg options for generating the svg shape.
+ * @returns {SVGGElement} the group element of the svg bubble.
+ */
+function createSvgShape ({ x, y, width, height, side, clazz }) {
```
`clazz` -> `class`

shubhi1092 commented on app/views/play/cinematic/DialogSystem.js:
```diff
@@ -0,0 +1,314 @@
+  return wrapWord(htmlString, / ?([^ ])+ ?/)
+}
+
+// Polyfill for `node.remove` method.
```
Can this polyfill be moved to a common place where app is initialized?

AndrewJakubowicz replied:
This is an amazing comment. Moving it below another polyfill in the `vendor.js` file.

shubhi1092 commented on app/views/play/cinematic/DialogSystem.js:
```diff
@@ -0,0 +1,314 @@
+   * @param {Object} context - used for interpolation.
+   */
+  set templateContext (context) {
+    this._context = context
```
Why are we not setting this in the constructor? Also is it generally called template context in the language of text interpolation, otherwise the name `context` doesnt look intuitive to me.

shubhi1092 replied:
I like `templateData`, or maybe `templateDataParameters` since parameter is the term used in their documentation.

shubhi1092 commented on app/views/play/cinematic/DialogSystem.js:
```diff
@@ -0,0 +1,314 @@
+ * @param {function} wrapWordString takes a word and can return a wrapped word.
+ * @returns {string} transformed html string.
+ */
+export function wrapText (htmlString, wrapLetterString, wrapWordString) {
```
Why are we exporting processText and wrapText, can we move them to a common place like `utils` if we want them to be re-usable?

AndrewJakubowicz replied:
I agree and really don't like these exports, so I'll give my problem hoping for a solution.

I wouldn't want to move these files to `utils` as they're very specific to the Dialog System. Specifically the `wrapText` function is very specific wrapping text in individual tags.
However these are also not easy functions so I want to ensure they have some test coverage.

A test that tests public behavior of the DialogSystem is difficult due to testing side effects. 
Thus I'm exporting these for testing to ensure their behavior is predictable for the system depending on them.

Interested in other tactics to solve the above problem.

shubhi1092 replied:
How about a common file thats specific to dialog system like `dialogSystemHelper`

AndrewJakubowicz replied:
hmmmm. The more I think about that the more I like it.
I see keeping the functions in `dialogSystem.js` as one option as they're very related.
But then they are exported muddying up with public exports of `dialogSystem`. Whilst placing them in a `dialogSystemHelper` file would allow them to be exported but it makes their intent clearer.
I'm a fan of the `dialogSystemHelper` idea 🚀 👏

AndrewJakubowicz replied:
This is also excellent because the `dialogsystem.js` file currently contains 2 ideas. The idea of drawing html nodes on a div canvas and the idea of drawing svgs onto an svg canvas.

These could both be split into separate files and the DialogSystem just uses them.
Not doing the full refactor in  this PR. But creating the initial folder and moving out the functions into the dialogSystemHelper file.

shubhi1092 commented on test/app/lib/cinematic/DialogSystem.spec.js:
```diff
@@ -0,0 +1,56 @@
+const dialogNode = text => ({ text })
+
+describe('dialogSystem', () => {
+  describe('text processing', () => {
```
it looks like this is only for testing wrapText, can modify description here to specify that

shubhi1092 commented on app/views/play/cinematic/DialogSystem.js:
```diff
@@ -0,0 +1,314 @@
+   * @param {import('../../../schemas/selectors/cinematic').DialogNode} dialogNode
+   * @returns {AbstractCommand[]}
+   */
+  parseDialogNode (dialogNode) {
```
Looks like we can also add test coverage for this method.

AndrewJakubowicz replied:
I agree but am still holding off as this method changes slightly again in the next  PR.
Your feedback is noted.

shubhi1092 commented:
Added some questions and suggestions. Otherwise looks good!

shubhi1092 commented:
I generally feel that the dialogSystem file is too big, and can probably be refactored.

AndrewJakubowicz commented:
@shubhi1092 I addressed your concerns. Some of them are addressed in the next PR with other changes.
Generally made the DialogSystem file smaller, which will encourage the files to stay smaller moving forward. That was an excellent suggestion.
